cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -427,8 +427,38 @@ void ShiftTime(FunctionContext* ctx, const CastOptions& options, const bool is_m
       out_data[i] = static_cast<out_type>(in_data[i]);
     }
   } else if (is_multiply) {
-    for (int64_t i = 0; i < input.length; i++) {
-      out_data[i] = static_cast<out_type>(in_data[i] * factor);
+    if (options.allow_time_truncate) {
```

Comment:
I reused `allow_time_truncate` for now, but it's not exactly the same, so should I add a `CastOptions.allow_time_overflow` or `allow_time_outofbounds` ?

Reply:
Hmm... unless we want to micro-optimize performance, I'm not sure why we would ever accept casting out-of-bounds. @xhochy @wesm what do you think about this?